lib/dal/src/workspace_snapshot/traits/attribute_value.rs

```diff
@@ -1,31 +1,175 @@
+        let mut sub_views = HashMap::new();
+        let mut forward_queue = VecDeque::from_iter([av_tree.root()]);
+        let mut work_stack = Vec::with_capacity(av_tree.count());
+
+        // Build the stack of work, ensuring that all children appear after their parent in the list.
+        while let Some(av_tree_entry) = forward_queue.pop_front() {
+            work_stack.push(av_tree_entry);
+
+            let children = av_tree.children_of(av_tree_entry.attribute_value_id);
+            forward_queue.reserve(children.len());
+            forward_queue.extend(children);
+        }
+
+        // Handle building the views in post-order traversal to ensure that all children have been
+        // processed before their parent.
+        while let Some(av_tree_entry) = work_stack.pop() {
+            let av_node_weight = self
+                .working_copy()
+                .await
+                .get_node_weight_by_id(av_tree_entry.attribute_value_id)?
+                .get_attribute_value_node_weight()?;
+            match av_node_weight.value() {
+                None => {
+                    // If the AttributeValue's value is None, and it does not have a component-specific
+                    // AttributePrototype, look for a default value for the Prop.
+                    if self
+                        .component_prototype_id(av_tree_entry.attribute_value_id)
+                        .await?
+                        .is_none()
+                    {
+                        if let Some(default_value) =
+                            self.prop_default_value(ctx, av_tree_entry.prop_id).await?
+                        {
+                            sub_views.insert(av_tree_entry.attribute_value_id, default_value);
+                        }
+                    }
+                }
+                Some(value_content_address) => {
+                    let value = match av_tree_entry.prop_kind {
+                        PropKind::Boolean
+                        | PropKind::Float
+                        | PropKind::Integer
+                        | PropKind::Json
+                        | PropKind::String => ctx
+                            .layer_db()
+                            .cas()
+                            .try_read_as::<si_events::CasValue>(
+                                &value_content_address.content_hash(),
+                            )
+                            .await?
+                            .map(Into::into)
+                            .ok_or_else(|| {
+                                WorkspaceSnapshotError::MissingContentFromStore(
+                                    av_tree_entry.attribute_value_id.into(),
+                                )
+                            })?,
+                        PropKind::Array => {
+                            let children = av_tree.children_of(av_tree_entry.attribute_value_id);
+                            let mut entries = Vec::with_capacity(children.len());
```

Comment:
Is this what we want here, given it will overallocate if the `if let Some()` is in the loop? (It might be what we want, especially if the map is likely to be close to the same size, because it means we don't have to do multiple allocations while growing. I'm not 100% clear what sub_views is for.)